File: Ejercicios_POO3.1_Java/src/Ejercicio2_Reservas_Hotel/Hotel.java
package Ejercicio2_Reservas_Hotel;


import java.util.ArrayList;
import java.util.List;
public class Hotel {
    private List<Habitacion> habitaciones = new ArrayList<>();
    private List<Reserva> reservas = new ArrayList<>();
    private List<Cliente> clientes = new ArrayList<>();

    public void agregarHabitacion(Habitacion habitacion) {
        habitaciones.add(habitacion);
    }

    public void registrarCliente(Cliente cliente) {
        clientes.add(cliente);
        System.out.println("Cliente registrado: " + cliente.getNombre() + " con ID " + cliente.getId());
    }

    public void registrarReserva(Cliente cliente, int numeroHabitacion) {
        Habitacion habitacion = buscarHabitacion(numeroHabitacion);
        if (habitacion != null && habitacion.isDisponible()) {
            habitacion.reservar();
            Reserva reserva = new Reserva(cliente, habitacion);
            reservas.add(reserva);
            System.out.println("Reserva realizada: " + reserva);
        } else {
            System.out.println("La habitación " + numeroHabitacion + " no está disponible.");
        }
    }

    public void cancelarReserva(Cliente cliente, int numeroHabitacion) {
        Reserva reserva = buscarReserva(cliente, numeroHabitacion);
        if (reserva != null) {
            reservas.remove(reserva);
            reserva.getHabitacion().liberar();
            System.out.println("Reserva cancelada para la habitación " + numeroHabitacion);
        } else {
            System.out.println("No se encontró una reserva para el cliente en la habitación " + numeroHabitacion);
        }
    }

    public void mostrarReservasActivas() {
        if (reservas.isEmpty()) {
            System.out.println("No hay reservas activas.");
        } else {
            System.out.println("Reservas activas:");
            for (Reserva reserva : reservas) {
                System.out.println(reserva);
            }
        }
    }

    public Habitacion buscarHabitacion(int numero) {
        for (Habitacion habitacion : habitaciones) {
            if (habitacion.getNumero() == numero) {
                return habitacion;
            }
        }
        return null;
    }

    public Reserva buscarReserva(Cliente cliente, int numeroHabitacion) {
        for (Reserva reserva : reservas) {
            if (reserva.getCliente().getId().equals(cliente.getId()) &&
                    reserva.getHabitacion().getNumero() == numeroHabitacion) {
                return reserva;
            }
        }
        return null;
    }

    public Cliente buscarClientePorId(String id) {
        for (Cliente cliente : clientes) {
            if (cliente.getId().equals(id)) {
                return cliente;
            }
        }
        return null;
    }
}

File: Ejercicios_POO3.1_Kotlin/src/Ejercicio2_Hotel/Hotel.kt
package Ejercicio2_Hotel


class Hotel {
    private val habitaciones = mutableListOf<Habitacion>()
    private val reservas = mutableListOf<Reserva>()
    private val clientes = mutableListOf<Cliente>()

    fun agregarHabitacion(habitacion: Habitacion) {
        habitaciones.add(habitacion)
    }

    fun registrarCliente(cliente: Cliente) {
        clientes.add(cliente)
        println("Cliente registrado: ${cliente.nombre} con ID ${cliente.id}")
    }

    fun registrarReserva(cliente: Cliente, numeroHabitacion: Int) {
        val habitacion = buscarHabitacion(numeroHabitacion)
        if (habitacion != null && habitacion.disponible) {
            habitacion.disponible = false
            val reserva = Reserva(cliente, habitacion)
            reservas.add(reserva)
            println("Reserva realizada: $reserva")
        } else {
            println("La habitación $numeroHabitacion no está disponible.")
        }
    }

    fun cancelarReserva(cliente: Cliente, numeroHabitacion: Int) {
        val reserva = buscarReserva(cliente, numeroHabitacion)
        if (reserva != null) {
            reservas.remove(reserva)
            reserva.habitacion.disponible = true
            println("Reserva cancelada para la habitación $numeroHabitacion")
        } else {
            println("No se encontró una reserva para el cliente en la habitación $numeroHabitacion")
        }
    }

    fun mostrarReservasActivas() {
        if (reservas.isEmpty()) {
            println("No hay reservas activas.")
        } else {
            println("Reservas activas:")
            reservas.forEach { println(it) }
        }
    }

    private fun buscarHabitacion(numero: Int): Habitacion? {
        return habitaciones.find { it.numero == numero }
    }

    private fun buscarReserva(cliente: Cliente, numeroHabitacion: Int): Reserva? {
        return reservas.find { it.cliente.id == cliente.id && it.habitacion.numero == numeroHabitacion }
    }

    fun buscarClientePorId(id: String): Cliente? {
        return clientes.find { it.id == id }
    }
}